nova_vm/src/ecmascript/builtins/control_abstraction_objects/promise_objects/promise_abstract_operations/promise_jobs.rs

```diff
@@ -0,0 +1,160 @@
+};
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct PromiseResolveThenableJob {
+    promise_to_resolve: Promise,
+    thenable: Object,
+    then: Function,
+}
+impl PromiseResolveThenableJob {
+    pub(crate) fn run(self, agent: &mut Agent) -> JsResult<()> {
+        // a. Let resolvingFunctions be CreateResolvingFunctions(promiseToResolve).
+        let promise_capability = PromiseCapability::from_promise(self.promise_to_resolve, false);
+        let resolve_function = agent
+            .heap
+            .create(PromiseResolvingFunctionHeapData {
+                object_index: None,
+                promise_capability,
+                resolve_type: PromiseResolvingFunctionType::Resolve,
+            })
+            .into_value();
+        let reject_function = agent
+            .heap
+            .create(PromiseResolvingFunctionHeapData {
+                object_index: None,
+                promise_capability,
+                resolve_type: PromiseResolvingFunctionType::Reject,
+            })
+            .into_value();
+
+        // b. Let thenCallResult be Completion(HostCallJobCallback(then, thenable, « resolvingFunctions.[[Resolve]], resolvingFunctions.[[Reject]] »)).
+        let then_call_result = call_function(
+            agent,
+            self.then,
+            self.thenable.into_value(),
+            Some(ArgumentsList(&[resolve_function, reject_function])),
+        );
+
+        // c. If thenCallResult is an abrupt completion, then
+        if let Err(err) = then_call_result {
+            // i. Return ? Call(resolvingFunctions.[[Reject]], undefined, « thenCallResult.[[Value]] »).
+            promise_capability.reject(agent, err.value());
+        }
+        // d. Return ? thenCallResult.
+        Ok(())
+    }
+}
+
+pub(crate) fn new_promise_resolve_thenable_job(
+    agent: &mut Agent,
+    promise_to_resolve: Promise,
+    thenable: Object,
+    then: Function,
+) -> Job {
+    // 2. Let getThenRealmResult be Completion(GetFunctionRealm(then.[[Callback]])).
+    // 5. NOTE: thenRealm is never null. When then.[[Callback]] is a revoked Proxy and no code runs, thenRealm is used to create error objects.
+    let then_realm = match get_function_realm(agent, then) {
+        // 3. If getThenRealmResult is a normal completion, let thenRealm be getThenRealmResult.[[Value]].
+        Ok(realm) => realm,
+        // 4. Else, let thenRealm be the current Realm Record.
+        Err(_) => agent.current_realm_id(),
+    };
+    // 6. Return the Record { [[Job]]: job, [[Realm]]: thenRealm }.
+    Job {
+        realm: Some(then_realm),
+        inner: InnerJob::PromiseResolveThenable(PromiseResolveThenableJob {
+            promise_to_resolve,
+            thenable,
+            then,
+        }),
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct PromiseReactionJob {
+    reaction: PromiseReaction,
+    argument: Value,
+}
+impl PromiseReactionJob {
+    pub(crate) fn run(self, agent: &mut Agent) -> JsResult<()> {
```

Comment:
**nitpick:** I'm quite missing many of the omitted spec steps and titles / links to the origin of the spec part.